lib/active_model/serializer/attributes.rb

```diff
@@ -0,0 +1,109 @@
+module ActiveModel
+  class Serializer
+    module Attributes
+      class Attribute
+        delegate :call, to: :reader
+
+        attr_reader :name, :reader
+
+        def initialize(name)
+          @name = name
+          @reader = :no_reader
+        end
+
+        def self.build(name, block)
+          if block
+            AttributeBlock.new(name, block)
+          else
+            AttributeReader.new(name)
+          end
+        end
+      end
+      class AttributeReader < Attribute
+        def initialize(name)
+          super(name)
+          @reader = ->(instance) { instance.read_attribute_for_serialization(name) }
+        end
+      end
+      class AttributeBlock < Attribute
+        def initialize(name, block)
+          super(name)
+          @reader = ->(instance) { instance.instance_eval(&block) }
+        end
+      end
```

Comment:
That would be fine by me, except I'd prefer not to add another class method, and without a reference to `name`, I think we'd have to go back to the old way of defining `_attributes_keys`, no?  All of which drive me to find polymorphic classes easier to reason about and use.

```diff
def attribute(name, options = {}, &block)
-      _attributes_keys[attr] = { key: key } if key != attr
end

+        # @api private
+        # maps attribute value to explict key name
+        # @see Serializer::attribute
+        # @see Adapter::FragmentCache#fragment_serializer
+        def _attributes_keys
+          _attribute_mappings
+            .each_with_object({}) do |(key, attribute_mapping), hash|
+              next if key == attribute_mapping.name
+              hash[attribute_mapping.name] = { key: key }
+            end
```